syntax = "proto3";

package splitbrain;

// NodeDef represents a node within the program graph.
//
// This can be be a file, BUILD target, or AST symbol. Note there is not enough
// information to split the program yet. Multi-file symbols are unsupported.
message NodeDef {
    // Unique name within the context of a single program.
    string name = 1;

    enum Kind {
        UNKNOWN = 0;
        FILE = 1;
        BUILD_TARGET = 2;
        SYMBOL = 3;
    }

    // The type of node. Is this a file or a AST symbol (e.g. function)?
    Kind kind = 2;

    // Name of out-edge symbols. For a program to be considered valid, all
    // symbolic references must be present in the graph.
    repeated string u_edge = 3;
}

// GraphDef represents a high-level program graph.
message GraphDef {
    repeated NodeDef symbol = 1;
}